fdbserver/storageserver.actor.cpp

```diff
@@ -1090,100 +1144,105 @@ ACTOR Future<Void> getValueQ( StorageServer* data, GetValueRequest req ) {
-				GetValueRequest getReq( span.context, req.key, latest, req.tags, req.debugID );
-				state Future<Void> getValue = getValueQ( data, getReq ); //we are relying on the delay zero at the top of getValueQ, if removed we need one here
-				GetValueReply reply = wait( getReq.reply.getFuture() );
-				span = Span(spanLocation, parent);
-				//TraceEvent("WatcherCheckValue").detail("Key",  req.key  ).detail("Value",  req.value  ).detail("CurrentValue",  v  ).detail("Ver", latest);
-
-				if(reply.error.present()) {
-					ASSERT(reply.error.get().code() != error_code_future_version);
-					throw reply.error.get();
-				}
-				if(BUGGIFY) {
-					throw transaction_too_old();
-				}
-				
-				DEBUG_MUTATION("ShardWatchValue", latest, MutationRef(MutationRef::DebugKey, req.key, reply.value.present() ? StringRef( reply.value.get() ) : LiteralStringRef("<null>") ) );
+	state Version minVersion = data->data().latestVersion;
+	state Future<Void> watchFuture = data->watches.onChange(metadata->key);
+	loop {
+		try {
+			metadata = data->getWatchMetadata(key);
+			state Version latest = data->version.get();
+			TEST(latest >= minVersion && latest < data->data().latestVersion); // Starting watch loop with latestVersion > data->version
+			GetValueRequest getReq( span.context, metadata->key, latest, metadata->tags, metadata->debugID );
+			state Future<Void> getValue = getValueQ( data, getReq ); //we are relying on the delay zero at the top of getValueQ, if removed we need one here
+			GetValueReply reply = wait( getReq.reply.getFuture() );
+			span = Span(spanLocation, parent);
+
+			if(reply.error.present()) {
+				ASSERT(reply.error.get().code() != error_code_future_version);
+				throw reply.error.get();
+			}
+			if(BUGGIFY) {
+				throw transaction_too_old();
+			}
+			
+			DEBUG_MUTATION("ShardWatchValue", latest, MutationRef(MutationRef::DebugKey, metadata->key, reply.value.present() ? StringRef( reply.value.get() ) : LiteralStringRef("<null>") ) );
 
-				if( req.debugID.present() )
-					g_traceBatch.addEvent("WatchValueDebug", req.debugID.get().first(), "watchValueQ.AfterRead"); //.detail("TaskID", g_network->getCurrentTask());
+			if( metadata->debugID.present() )
+				g_traceBatch.addEvent("WatchValueDebug", metadata->debugID.get().first(), "watchValueQ.AfterRead"); //.detail("TaskID", g_network->getCurrentTask());
 
-				if( reply.value != req.value ) {
-					req.reply.send(WatchValueReply{ latest });
-					return Void();
-				}
+			if( reply.value != metadata->value && latest >= metadata->version ) {
+				return latest; // fire watch
+			}
 
-				if( data->watchBytes > SERVER_KNOBS->MAX_STORAGE_SERVER_WATCH_BYTES ) {
-					TEST(true); //Too many watches, reverting to polling
-					data->sendErrorWithPenalty(req.reply, watch_cancelled(), data->getPenalty());
-					return Void();
-				}
+			if( data->watchBytes > SERVER_KNOBS->MAX_STORAGE_SERVER_WATCH_BYTES ) {
+				TEST(true); //Too many watches, reverting to polling
```

Comment:
We need to check that we still run into this. Can you please check in Snowhouse whether you tests triggered this? If not we can buggify this knob to be even smaller (currently the buggify value is 10KB which might be just small enough).

Reply:
Yes I validated that 10K is indeed small enough to trigger the too many watches exception and revert to polling